add language switcher panel

BernhardBaumrock opened the pull request:
![image](https://user-images.githubusercontent.com/8488586/178543365-149d971a-3d82-4b10-abe4-c47c533cd11e.png)

One of the few things that I really missed from AdminOnSteroids - so I built a tracy panel instead :) I mean AOS has many great features but only a few that I really need for my daily work. This is one of them.

BernhardBaumrock commented:
The icon color seems a bit off the others! Don't know how your css works so you might want to change that yourself, thx :)

teppokoivula commented:
> The icon color seems a bit off the others! Don't know how your css works so you might want to change that yourself, thx :)

This is a bit off-topic (sorry!) but for the Wireframe panel I ended up setting fill color as hex code as well. It would be nice if currentcolor worked here, just in case that Tracy styles are ever modified. Please let me know if this is already possible... :)

adrianbj commented:
Thank you @BernhardBaumrock - great idea! 

I just have a few comments if you would mind addressing:

1. For consistency with the rest of the code please convert `$this->wire->input` etc into `$this->wire('input')`. Perhaps at some point I'll globally replace to Ryan's current favorite `$this->wire()->input`, but for now, I'd like to be as consistent as possible.
2. I don't think there is any real reason for the maximize button for this panel - you'll notice I don't add this to all - only the ones where it makes sense.
3. I'd love to see the color of the icon change to the yellow warning when the language has been switched.
4. Regarding icon colors, please use `TracyDebugger::COLOR_NORMAL` and `TracyDebugger::COLOR_WARN` (https://github.com/adrianbj/TracyDebugger/blob/d736e903eb9a0f95f4487d4ba451903475e373ce/TracyDebugger.module.php#L44-L46)
5. Are there versions of https://github.com/adrianbj/TracyDebugger/pull/77/files#diff-d31bb28a16012b51d4f692aef359c04bb4533ef077d01441f1e25e5e8c56e838R1450-R1451 for other admin themes? Maybe not relevant anymore (hopefully everything is just skinning UiKit with your modules), but just thought I'd ask.

adrianbj commented:
@teppokoivula - do those color constants take care of your needs?

adrianbj commented:
Sorry, another idea / request - I feel like the language selection needs to be a `<select>` because it has the potential to be a very long list on some site.

BernhardBaumrock commented:
> For consistency with the rest of the code please convert $this->wire->input etc into $this->wire('input'). Perhaps at some point I'll globally replace to Ryan's current favorite $this->wire()->input, but for now, I'd like to be as consistent as possible.

Hey @adrianbj I've added a commit that replaces `$this->wire` with `$this->wire()` but I just realized that is not the syntax you wanted.

Though I have to vote against your syntax as it breaks proper intellisense:

Using your suggested syntax the IDE does not know that $this->wire('user') is a user object. It just uses all possible return types for suggestions which in this case also suggests methods of $this->wire->files:

![image](https://user-images.githubusercontent.com/8488586/178706482-2694ebf8-6c0e-4171-ac50-367580f0b184.png)

Using my preferred syntax the suggestions are correct:

![image](https://user-images.githubusercontent.com/8488586/178707022-0bcc1615-794d-4890-920d-197388d2b06c.png)

I know that it has a little overhead using magic properties instead of direct method calls, but I think that is really not worth taking into account here and giving up proper intellisense.

So imho the best would be `$this->wire()->user()...` but that's a lot of useless brackets in the code...

BernhardBaumrock commented:
@adrianbj any more thoughts on my PR? :) I'm quite sure it's not perfect yet, but it's usable already and would be helpful for my projects

adrianbj commented:
Hi @BernhardBaumrock - sorry, I had a few changes I wanted to make and didn't have time until now.

1. Change to multiline select / options.
2. Highlight language when changed vs default.
3. Add "Reset" button so you don't need to find your default language to switch back to it.
4. Remove need for `getFresh()` so Tracy can still support old versions of PW.
5. Fix fatal exception when setting the language (didn't see locally, but appeared on one of my live servers)

Please let me know if you have any problems with this new version.

BernhardBaumrock commented:
Thx @adrianbj for merging :) 

I've found one little issue: When using the new "reset" the backend menu cache is not flushed and stays in the previously selected language. When switching via the select field it works as expected.

Personally I'd prefer if the sort order of the languages did not change depending on which language is selected.

> Change to multiline select / options.

Ok... I've added a maxheight to my version so that it does not make problems. I don't see the benefit of having a select, because now I need 2 clicks to change the language instead of one. But it's of course fine to make your module work the way you prefer it.

> Highlight language when changed vs default.

I've already implemented that as well, so I'm not sure why you are listing that as a change?

> Add "Reset" button so you don't need to find your default language to switch back to it.

Good idea!

> Remove need for getFresh() so Tracy can still support old versions of PW.

Thx!

> Fix fatal exception when setting the language (didn't see locally, but appeared on one of my live servers)

Ok, thx, never seen such an exception on my end :)

adrianbj commented:
> I've found one little issue: When using the new "reset" the backend menu cache is not flushed

Thanks - should be fixed now.

> Personally I'd prefer if the sort order of the languages did not change depending on which language is selected.

I've changed this now - the key thing was ensuring that the selected language was visible so now I am using JS to scroll it into view which is important when a site has a lot of languages.

> now I need 2 clicks to change the language instead of one

Ok, the selection now auto-submits the form.

> I've already implemented that as well, so I'm not sure why you are listing that as a change?

The key difference is the highlighting of the default language when there hasn't been a switch.

Better now?

BernhardBaumrock commented:
hi @adrianbj thx, everything is great, sorry for the details, would have also been great before ;)

adrianbj commented:
Details are good :)

teppokoivula commented:
> @teppokoivula - do those color constants take care of your needs?

Thanks — [works great](https://github.com/wireframe-framework/Wireframe/commit/30b57eb935baea38ec983333f0162d7aab3aa49f) :)
